File: android/app/src/main/java/cz/drabek/chuvicka/parent/Onvif.kt
package cz.drabek.chuvicka.parent

import cz.drabek.chuvicka.Log
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.security.SecureRandom
import java.text.SimpleDateFormat
import java.util.Base64
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/** The four ways to turn the camera. */
enum class PtzDirection { UP, DOWN, LEFT, RIGHT }

/**
 * ONVIF PTZ, the standard way to turn an IP camera: SOAP 1.2 over HTTP with a WS-Security login
 * (UsernameToken, PasswordDigest). Only what the app needs: GetProfiles, ContinuousMove, Stop.
 * Pure functions, no android.*, so the unit tests run them.
 */
object Onvif {
    private const val WSSE = "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd"
    private const val WSU = "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd"
    private const val DIGEST_TYPE = "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-username-token-profile-1.0#PasswordDigest"
    private const val PTZ_NS = "xmlns:tptz=\"http://www.onvif.org/ver20/ptz/wsdl\" xmlns:tt=\"http://www.onvif.org/ver10/schema\""

    /** The WS-Security header. Digest = Base64(SHA-1(nonce + created + password)). */
    fun securityHeader(user: String, password: String, nonce: ByteArray, created: String): String {
        val sha = MessageDigest.getInstance("SHA-1")
        sha.update(nonce)
        sha.update(created.toByteArray(Charsets.UTF_8))
        sha.update(password.toByteArray(Charsets.UTF_8))
        val digest = Base64.getEncoder().encodeToString(sha.digest())
        val nonce64 = Base64.getEncoder().encodeToString(nonce)
        return "<wsse:Security xmlns:wsse=\"$WSSE\" xmlns:wsu=\"$WSU\"><wsse:UsernameToken>" +
            "<wsse:Username>${xml(user)}</wsse:Username>" +
            "<wsse:Password Type=\"$DIGEST_TYPE\">$digest</wsse:Password>" +
            "<wsse:Nonce>$nonce64</wsse:Nonce><wsu:Created>$created</wsu:Created>" +
            "</wsse:UsernameToken></wsse:Security>"
    }

    fun envelope(body: String, security: String): String =
        "<?xml version=\"1.0\"?><s:Envelope xmlns:s=\"http://www.w3.org/2003/05/soap-envelope\">" +
            "<s:Header>$security</s:Header><s:Body>$body</s:Body></s:Envelope>"

    const val getProfilesBody = "<trt:GetProfiles xmlns:trt=\"http://www.onvif.org/ver10/media/wsdl\"/>"

    fun continuousMoveBody(token: String, x: Float, y: Float): String =
        "<tptz:ContinuousMove $PTZ_NS><tptz:ProfileToken>${xml(token)}</tptz:ProfileToken>" +
            "<tptz:Velocity><tt:PanTilt x=\"${number(x)}\" y=\"${number(y)}\"/></tptz:Velocity></tptz:ContinuousMove>"

    fun stopBody(token: String): String =
        "<tptz:Stop $PTZ_NS><tptz:ProfileToken>${xml(token)}</tptz:ProfileToken>" +
            "<tptz:PanTilt>true</tptz:PanTilt></tptz:Stop>"

    private val profileRegex = Regex("<(?:\\w+:)?Profiles\\b([^>]*)>(.*?)</(?:\\w+:)?Profiles>", RegexOption.DOT_MATCHES_ALL)
    private val tokenRegex = Regex("\\btoken\\s*=\\s*[\"']([^\"']*)[\"']")

    /** The token of the first profile that can turn (it has a PTZConfiguration), from a GetProfiles answer. */
    fun ptzProfileToken(xml: String): String? {
        for (m in profileRegex.findAll(xml)) {
            if (!m.groupValues[2].contains("PTZConfiguration")) continue
            val token = tokenRegex.find(m.groupValues[1])?.groupValues?.get(1)
            if (!token.isNullOrEmpty()) return token
        }
        return null
    }

    /** The time for the login, in UTC. */
    fun createdNow(): String =
        SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply { timeZone = TimeZone.getTimeZone("UTC") }.format(Date())

    private fun number(v: Float) = String.format(Locale.US, "%.1f", v)

    private fun xml(s: String) = s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
        .replace("\"", "&quot;").replace("'", "&apos;")
}

/**
 * The ONVIF connection to one camera. Blocking: call it on Dispatchers.IO or the monitor thread.
 * Same login as RTSP. It keeps the PTZ profile token in memory only.
 */
class OnvifClient(private val host: String, private val port: Int,
                  private val user: String, private val password: String) {
    @Volatile private var token: String? = null
    private val random = SecureRandom()

    /** The PTZ profile token, or null when the camera cannot turn (or does not answer). */
    fun loadPtzProfile(): String? {
        token?.let { return it }
        val (code, body) = post(Onvif.getProfilesBody) ?: return null
        if (code != 200) { Log.add("camera control: GetProfiles answered HTTP $code"); return null }
        val t = Onvif.ptzProfileToken(body)
        if (t == null) Log.add("camera control: no profile that can turn")
        token = t
        return t
    }

    fun move(x: Float, y: Float): Boolean {
        val t = loadPtzProfile() ?: return false
        return ok("ContinuousMove", post(Onvif.continuousMoveBody(t, x, y)))
    }

    fun stop(): Boolean {
        val t = loadPtzProfile() ?: return false
        return ok("Stop", post(Onvif.stopBody(t)))
    }

    /** One step: move, wait, stop. Stop also after a failed move, so the camera never keeps turning. */
    @Synchronized
    fun step(x: Float, y: Float, seconds: Float = 0.6f): Boolean {
        val moved = move(x, y)
        if (moved) {
            try { Thread.sleep((seconds * 1000).toLong()) } catch (_: InterruptedException) {}
        }
        val stopped = stop()
        return moved && stopped
    }

    private fun ok(what: String, answer: Pair<Int, String>?): Boolean {
        val code = answer?.first ?: return false
        if (code != 200) Log.add("camera control: $what answered HTTP $code")
        return code == 200
    }

    /** One SOAP request: the HTTP status and the body, or null when the camera does not answer. */
    private fun post(body: String): Pair<Int, String>? {
        val nonce = ByteArray(16).also { random.nextBytes(it) }
        val xml = Onvif.envelope(body, Onvif.securityHeader(user, password, nonce, Onvif.createdNow()))
        var c: HttpURLConnection? = null
        return try {
            c = URL("http://$host:$port/onvif/service").openConnection() as HttpURLConnection
            c.requestMethod = "POST"
            c.doOutput = true
            c.connectTimeout = 4000
            c.readTimeout = 4000
            c.setRequestProperty("Content-Type", "application/soap+xml; charset=utf-8")
            c.outputStream.use { it.write(xml.toByteArray(Charsets.UTF_8)) }
            val code = c.responseCode
            val stream = if (code in 200..299) c.inputStream else c.errorStream
            val text = stream?.use { String(it.readBytes(), Charsets.UTF_8) } ?: ""
            code to text
        } catch (e: IOException) {
            Log.add("camera control: $host:$port does not answer (${e.javaClass.simpleName})")
            null
        } finally {
            c?.disconnect()
        }
    }
}

File: Nursery/Stream/ONVIF.swift
import CryptoKit
import Foundation

/// ONVIF pan and tilt: SOAP 1.2 over HTTP, with a WS-Security UsernameToken (PasswordDigest).
/// The camera account is the same as for RTSP. The functions here are pure, for the tests;
/// `ONVIFClient` sends them.
enum ONVIF {
    static let getProfilesBody = #"<trt:GetProfiles xmlns:trt="http://www.onvif.org/ver10/media/wsdl"/>"#

    /// The WS-Security header. digest = base64(SHA1(nonce + created + password)).
    static func securityHeader(user: String, password: String, nonce: [UInt8], created: String) -> String {
        var data = Data(nonce)
        data.append(Data(created.utf8))
        data.append(Data(password.utf8))
        let digest = Data(Insecure.SHA1.hash(data: data)).base64EncodedString()
        let nonce64 = Data(nonce).base64EncodedString()
        return #"<wsse:Security xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">"#
            + "<wsse:UsernameToken><wsse:Username>\(escape(user))</wsse:Username>"
            + #"<wsse:Password Type="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-username-token-profile-1.0#PasswordDigest">"#
            + "\(digest)</wsse:Password><wsse:Nonce>\(nonce64)</wsse:Nonce><wsu:Created>\(created)</wsu:Created>"
            + "</wsse:UsernameToken></wsse:Security>"
    }

    static func envelope(body: String, security: String) -> String {
        #"<?xml version="1.0"?><s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope"><s:Header>"#
            + security + "</s:Header><s:Body>" + body + "</s:Body></s:Envelope>"
    }

    /// Turn at this speed until Stop. x: left −, right +. y: down −, up +.
    static func continuousMoveBody(token: String, x: Float, y: Float) -> String {
        let vx = String(format: "%.1f", Double(x))
        let vy = String(format: "%.1f", Double(y))
        return #"<tptz:ContinuousMove xmlns:tptz="http://www.onvif.org/ver20/ptz/wsdl" xmlns:tt="http://www.onvif.org/ver10/schema">"#
            + "<tptz:ProfileToken>\(escape(token))</tptz:ProfileToken><tptz:Velocity>"
            + "<tt:PanTilt x=\"\(vx)\" y=\"\(vy)\"/>"
            + "</tptz:Velocity></tptz:ContinuousMove>"
    }

    static func stopBody(token: String) -> String {
        #"<tptz:Stop xmlns:tptz="http://www.onvif.org/ver20/ptz/wsdl" xmlns:tt="http://www.onvif.org/ver10/schema">"#
            + "<tptz:ProfileToken>\(escape(token))</tptz:ProfileToken><tptz:PanTilt>true</tptz:PanTilt></tptz:Stop>"
    }

    private static let profilePattern = try! NSRegularExpression(
        pattern: #"<(?:[A-Za-z0-9_.-]+:)?Profiles\b([^>]*)>(.*?)</(?:[A-Za-z0-9_.-]+:)?Profiles\s*>"#,
        options: [.dotMatchesLineSeparators])
    private static let tokenPattern = try! NSRegularExpression(pattern: #"\btoken\s*=\s*(?:"([^"]*)"|'([^']*)')"#)

    /// The token of the first profile that can turn (it has a PTZConfiguration), from the
    /// GetProfiles answer. The token is not the first attribute (`fixed="true" token="profile_1"`).
    static func ptzProfileToken(in xml: String) -> String? {
        let ns = xml as NSString
        for m in profilePattern.matches(in: xml, range: NSRange(location: 0, length: ns.length)) {
            let body = ns.substring(with: m.range(at: 2))
            guard body.contains("PTZConfiguration") else { continue }
            let attributes = ns.substring(with: m.range(at: 1))
            let a = attributes as NSString
            guard let t = tokenPattern.firstMatch(in: attributes, range: NSRange(location: 0, length: a.length)) else { continue }
            let value = t.range(at: 1).location != NSNotFound ? a.substring(with: t.range(at: 1)) : a.substring(with: t.range(at: 2))
            if !value.isEmpty { return value }
        }
        return nil
    }

    /// The time for `Created`: UTC, with no fraction.
    static func createdNow() -> String {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        return f.string(from: Date())
    }

    private static func escape(_ s: String) -> String {
        s.replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
            .replacingOccurrences(of: "\"", with: "&quot;")
            .replacingOccurrences(of: "'", with: "&apos;")
    }
}

/// It turns one camera over ONVIF. It lives on the main actor: the state is small, and each
/// request waits off the main thread in URLSession.
@MainActor
final class ONVIFClient {
    let host: String
    let port: UInt16
    let user: String
    private let password: String
    /// The PTZ profile, in memory only. A new client asks again.
    private var token: String?
    /// One step at a time: a held arrow repeats faster than a step ends.
    private var stepping = false
    private let session: URLSession = {
        let c = URLSessionConfiguration.ephemeral
        c.timeoutIntervalForRequest = 4
        c.waitsForConnectivity = false
        return URLSession(configuration: c)
    }()

    init(host: String, port: UInt16, user: String, password: String) {
        self.host = host
        self.port = port
        self.user = user
        self.password = password
    }

    /// GetProfiles, then the token of the first profile that can turn. Nil: the camera cannot turn,
    /// or it does not answer.
    func loadPTZProfile() async -> String? {
        if let token { return token }
        guard let xml = await call(ONVIF.getProfilesBody, what: "GetProfiles") else { return nil }
        token = ONVIF.ptzProfileToken(in: xml)
        if token == nil { Log.shared.add("ONVIF: no PTZ profile on \(host):\(port)") }
        return token
    }

    func move(x: Float, y: Float) async -> Bool {
        guard let token = await loadPTZProfile() else { return false }
        return await call(ONVIF.continuousMoveBody(token: token, x: x, y: y), what: "ContinuousMove") != nil
    }

    func stop() async -> Bool {
        guard let token = await loadPTZProfile() else { return false }
        return await call(ONVIF.stopBody(token: token), what: "Stop") != nil
    }

    /// One step: turn, wait, stop. Stop also after a failed move, so the camera never keeps turning.
    /// A step that comes while one runs is skipped: the camera still turns from the running one.
    func step(x: Float, y: Float, seconds: Double = 0.6) async -> Bool {
        guard !stepping else { return true }
        stepping = true
        defer { stepping = false }
        let moved = await move(x: x, y: y)
        if moved { try? await Task.sleep(for: .seconds(seconds)) }
        let stopped = await stop()
        return moved && stopped
    }

    /// One SOAP request. The answer text on HTTP 200, else nil. The log line has no password.
    private func call(_ body: String, what: String) async -> String? {
        guard let url = URL(string: "http://\(host):\(port)/onvif/service") else { return nil }
        let nonce = (0..<16).map { _ in UInt8.random(in: 0...255) }
        let security = ONVIF.securityHeader(user: user, password: password, nonce: nonce, created: ONVIF.createdNow())
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/soap+xml; charset=utf-8", forHTTPHeaderField: "Content-Type")
        req.httpBody = Data(ONVIF.envelope(body: body, security: security).utf8)
        do {
            let (data, response) = try await session.data(for: req)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            guard status == 200 else {
                // 400 with a SOAP Fault: usually a wrong user or password.
                Log.shared.add("ONVIF \(what) on \(host):\(port): HTTP \(status)")
                return nil
            }
            return String(decoding: data, as: UTF8.self)
        } catch {
            Log.shared.add("ONVIF \(what) on \(host):\(port) failed: \(error.localizedDescription)")
            return nil
        }
    }
}
